File: testgen/ui/views/overview.py
import typing
from datetime import datetime

import pandas as pd
import streamlit as st
from pandas.api.types import is_string_dtype

import testgen.ui.services.database_service as db
from testgen.common import date_service
from testgen.ui.components import widgets as testgen
from testgen.ui.navigation.menu import MenuItem
from testgen.ui.navigation.page import Page
from testgen.ui.queries import project_queries
from testgen.ui.services import test_suite_service
from testgen.ui.session import session
from testgen.utils import to_int, truncate

STALE_PROFILE_DAYS = 30
PAGE_ICON = "home"


class OverviewPage(Page):
    path = "overview"
    can_activate: typing.ClassVar = [
        lambda: session.authentication_status,
    ]
    menu_item = MenuItem(icon=PAGE_ICON, label="Overview", order=0)

    def render(self, project_code: str | None = None, **_kwargs):
        testgen.page_header(
            "Project Overview",
            "introduction-to-dataops-testgen",
        )

        project_code = project_code or session.project
        table_groups_df: pd.DataFrame = get_table_groups_summary(project_code)
        render_project_summary(table_groups_df)

        if render_empty_state(project_code):
            return

        table_group_header_col, table_group_filter_col, table_group_sort_col = st.columns([0.6, 0.2, 0.2])
        table_group_header_col.html(f'<h5 style="margin-top: 16px;">Table Groups ({len(table_groups_df.index)})</h5>')
        with table_group_filter_col:
            name_filter = st.text_input(label="Search by table group name")
            table_groups_df = table_groups_df.loc[
                table_groups_df["table_groups_name"].str.contains(name_filter, case=False)
            ]

        with table_group_sort_col:
            table_groups_df["latest_activity_date"] = table_groups_df[
                ["latest_profile_start", "latest_tests_start"]
            ].apply(pd.to_datetime).max(axis=1) # apply is needed to handle missing values
            ascending_fields: list[str] = ["table_groups_name"]
            sort_options  = pd.DataFrame({
                "value": ["table_groups_name", "latest_activity_date"],
                "label": ["Table group name", "Latest activity"],
            })

            sort_by = testgen.select(
                label="Sort by",
                options=sort_options,
                required=True,
                default_value="latest_activity_date",
                display_column="label",
                value_column="value",
            )

            table_groups_df.sort_values(
                by=typing.cast(str, sort_by),
                ascending=sort_by in ascending_fields,
                inplace=True,
                key=lambda column: column.str.lower() if is_string_dtype(column) else column,
            )

        for index, table_group in table_groups_df.iterrows():
            render_table_group_card(table_group, project_code, index)


def render_empty_state(project_code: str) -> bool:
    project_summary_df = project_queries.get_summary_by_code(project_code)
    if project_summary_df["profiling_runs_ct"] or project_summary_df["test_runs_ct"]:
        return False

    label="Your project is empty"
    testgen.whitespace(3)
    if not project_summary_df["connections_ct"]:
        testgen.empty_state(
            label=label,
            icon=PAGE_ICON,
            message=testgen.EmptyStateMessage.Connection,
            action_label="Go to Connections",
            link_href="connections",
        )
    else:
        testgen.empty_state(
            label=label,
            icon=PAGE_ICON,
            message=testgen.EmptyStateMessage.Profiling if project_summary_df["table_groups_ct"] else testgen.EmptyStateMessage.TableGroup,
            action_label="Go to Table Groups",
            link_href="connections:table-groups",
            link_params={ "connection_id": str(project_summary_df["default_connection_id"]) }
        )
    return True


def render_project_summary(table_groups: pd.DataFrame) -> None:
    project_column, _ = st.columns([.5, .5])
    with project_column:
        testgen.whitespace(0.3)
        with testgen.card():
            summary_column, _ = st.columns([.8, .2])
            # TODO: Uncomment and replace with below section when adding the score
            # score_column, summary_column = st.columns([.5, .5])

            # with score_column:
            #     st.caption("Project HIT score")
            #     st.metric(
            #         "Project HIT score",
            #         value=project_score,
            #         delta=project_score_delta or 0,
            #         label_visibility="collapsed",
            #     )

            with summary_column:
                st.caption("Project Summary")
                st.html(f"""<b>{len(table_groups.index)}</b> table groups
                            <br><b>{int(table_groups['latest_tests_suite_ct'].sum())}</b> test suites
                            <br><b>{int(table_groups['latest_tests_ct'].sum())}</b> test definitions
                            """)


@st.fragment
def render_table_group_card(table_group: pd.Series, project_code: str, key: int) -> None:
    with testgen.card(title=table_group["table_groups_name"]) as test_suite_card:

        # Don't remove this
        # For some reason, st.columns do not get completely removed from DOM when used conditionally within a fragment
        # Without this CSS, the "hidden" elements in the expanded state take up space
        testgen.no_flex_gap()

        with test_suite_card.actions:
            expand_toggle = testgen.expander_toggle(key=f"toggle_{key}")

        profile_column, tests_column = st.columns([.5, .5])
        # TODO: Uncomment and replace with below section when adding the score
        # score_column, profile_column, tests_column = st.columns([.2, .35, .45])

        # with score_column:
        #     st.caption("HIT score")
        #     st.metric(
        #         "HIT score",
        #         value=table_group["score"],
        #         delta=table_group["score_delta"] or 0,
        #         label_visibility="collapsed",
        #     )

        with profile_column:
            testgen.no_flex_gap()
            latest_profile_start = table_group["latest_profile_start"]

            stale_message = ""
            if pd.notnull(latest_profile_start) and (profile_days_ago := (datetime.utcnow() - latest_profile_start).days) > STALE_PROFILE_DAYS:
                stale_message = f'<span style="color: var(--red);">({profile_days_ago} days ago)</span>'
            testgen.caption(f"Latest profile {stale_message}")

            if pd.notnull(latest_profile_start):
                testgen.link(
                    label=date_service.get_timezoned_timestamp(st.session_state, latest_profile_start),
                    href="profiling-runs:results",
                    params={ "run_id": str(table_group["latest_profile_id"]) },
                    key=f"overview:keys:go-to-profile:{table_group['latest_profile_id']}",
                )

                anomaly_count = to_int(table_group["latest_anomalies_ct"])
                with st.container():
                    testgen.flex_row_start()
                    testgen.text(f"""
                                 <b>{to_int(table_group['latest_profile_table_ct'])}</b> tables &nbsp;|&nbsp;
                                 <b>{to_int(table_group['latest_profile_column_ct'])}</b> columns &nbsp;|
                                 """)
                    testgen.link(
                        label=f"{anomaly_count} hygiene issues",
                        href="profiling-runs:hygiene",
                        params={ "run_id": str(table_group["latest_profile_id"]) },
                        width=150,
                        key=f"overview:keys:go-to-issues:{table_group['latest_profile_id']}",
                    )

                if anomaly_count:
                    testgen.summary_bar(
                        items=[
                            { "label": "Definite", "value": to_int(table_group["latest_anomalies_definite_ct"]), "color": "red" },
                            { "label": "Likely", "value": to_int(table_group["latest_anomalies_likely_ct"]), "color": "orange" },
                            { "label": "Possible", "value": to_int(table_group["latest_anomalies_possible_ct"]), "color": "yellow" },
                            { "label": "Dismissed", "value": to_int(table_group["latest_anomalies_dismissed_ct"]), "color": "grey" },
                        ],
                        height=12,
                        width=280,
                    )
            else:
                st.markdown("--")

        with tests_column:
            testgen.no_flex_gap()
            st.caption("Latest test results")
            total_tests = to_int(table_group["latest_tests_ct"])
            if total_tests:
                passed_tests = to_int(table_group["latest_tests_passed_ct"])
                testgen.text(f"{truncate(passed_tests * 100 / total_tests)}% passed")
                testgen.text(f"<b>{total_tests}</b> tests in <b>{to_int(table_group['latest_tests_suite_ct'])}</b> test suites", "margin: 12px 0 12px;")

                testgen.summary_bar(
                items=[
                    { "label": "Passed", "value": passed_tests, "color": "green" },
                    { "label": "Warning", "value": to_int(table_group["latest_tests_warning_ct"]), "color": "yellow" },
                    { "label": "Failed", "value": to_int(table_group["latest_tests_failed_ct"]), "color": "red" },
                    { "label": "Error", "value": to_int(table_group["latest_tests_error_ct"]), "color": "brown" },
                    { "label": "Dismissed", "value": to_int(table_group["latest_tests_dismissed_ct"]), "color": "grey" },
                ],
                height=12,
                width=350,
            )
            else:
                st.markdown("--")

        if expand_toggle:
            render_table_group_expanded(table_group["id"], project_code)


def render_table_group_expanded(table_group_id: str, project_code: str) -> None:
    testgen.divider(8, 12)

    column_spec = [0.25, 0.15, 0.15, 0.5]
    suite_column, generation_column, run_column, results_column = st.columns(column_spec)
    suite_column.caption("Test Suite")
    generation_column.caption("Latest Generation")
    run_column.caption("Latest Run")
    results_column.caption("Latest Results")
    testgen.whitespace(1)

    test_suites_df: pd.DataFrame = test_suite_service.get_by_project(project_code, table_group_id)

    for _, suite in test_suites_df.iterrows():
        render_test_suite_item(suite, column_spec)


def render_test_suite_item(test_suite: pd.Series, column_spec: list[int]) -> None:
    suite_column, generation_column, run_column, results_column = st.columns(column_spec)
    with suite_column:
        testgen.no_flex_gap()
        testgen.link(
            label=test_suite["test_suite"],
            href="test-suites:definitions",
            params={ "test_suite_id": str(test_suite["id"]) },
            key=f"overview:keys:go-to-definitions:{test_suite['id']}",
        )
        testgen.caption(f"{to_int(test_suite['test_ct'])} tests", "margin-top: -16px;")

    with generation_column:
        if (latest_generation := test_suite["latest_auto_gen_date"]) and pd.notnull(latest_generation):
            testgen.text(date_service.get_timezoned_timestamp(st.session_state, latest_generation))
        else:
            st.markdown("--")

    with run_column:
        if (latest_run_start := test_suite["latest_run_start"]) and pd.notnull(latest_run_start):
            testgen.link(
                label=date_service.get_timezoned_timestamp(st.session_state, latest_run_start),
                href="test-runs:results",
                params={ "run_id": str(test_suite["latest_run_id"]) },
                key=f"overview:keys:go-to-run:{test_suite['latest_run_id']}",
            )
        else:
            st.markdown("--")

    with results_column:
        if to_int(test_suite["last_run_test_ct"]):
            testgen.summary_bar(
                items=[
                    { "label": "Passed", "value": to_int(test_suite["last_run_passed_ct"]), "color": "green" },
                    { "label": "Warning", "value": to_int(test_suite["last_run_warning_ct"]), "color": "yellow" },
                    { "label": "Failed", "value": to_int(test_suite["last_run_failed_ct"]), "color": "red" },
                    { "label": "Error", "value": to_int(test_suite["last_run_error_ct"]), "color": "brown" },
                    { "label": "Dismissed", "value": to_int(test_suite["last_run_dismissed_ct"]), "color": "grey" },
                ],
                height=8,
                width=200,
            )
        else:
            st.markdown("--")


def get_table_groups_summary(project_code: str) -> pd.DataFrame:
    schema = st.session_state["dbschema"]
    sql = f"""
    WITH latest_profile AS (
        SELECT latest_run.table_groups_id,
            latest_run.id,
            latest_run.profiling_starttime,
            latest_run.table_ct,
            latest_run.column_ct,
            latest_run.anomaly_ct,
            SUM(
                CASE
                    WHEN COALESCE(latest_anomalies.disposition, 'Confirmed') = 'Confirmed'
                    AND anomaly_types.issue_likelihood = 'Definite' THEN 1
                    ELSE 0
                END
            ) as definite_ct,
            SUM(
                CASE
                    WHEN COALESCE(latest_anomalies.disposition, 'Confirmed') = 'Confirmed'
                    AND anomaly_types.issue_likelihood = 'Likely' THEN 1
                    ELSE 0
                END
            ) as likely_ct,
            SUM(
                CASE
                    WHEN COALESCE(latest_anomalies.disposition, 'Confirmed') = 'Confirmed'
                    AND anomaly_types.issue_likelihood = 'Possible' THEN 1
                    ELSE 0
                END
            ) as possible_ct,
            SUM(
                CASE
                    WHEN COALESCE(latest_anomalies.disposition, 'Confirmed') IN ('Dismissed', 'Inactive')
                    AND anomaly_types.issue_likelihood <> 'Potential PII' THEN 1
                    ELSE 0
                END
            ) as dismissed_ct
        FROM {schema}.table_groups groups
            LEFT JOIN {schema}.profiling_runs latest_run ON (
                groups.last_complete_profile_run_id = latest_run.id
            )
            LEFT JOIN {schema}.profile_anomaly_results latest_anomalies ON (
                latest_run.id = latest_anomalies.profile_run_id
            )
            LEFT JOIN {schema}.profile_anomaly_types anomaly_types ON (
                anomaly_types.id = latest_anomalies.anomaly_id
            )
        GROUP BY latest_run.id
    ),
    latest_tests AS (
        SELECT suites.table_groups_id,
            MAX(latest_run.test_starttime) AS test_starttime,
            COUNT(DISTINCT latest_run.test_suite_id) as test_suite_ct,
            COUNT(latest_results.id) as test_ct,
            SUM(
                CASE
                    WHEN COALESCE(latest_results.disposition, 'Confirmed') = 'Confirmed'
                    AND latest_results.result_status = 'Passed' THEN 1
                    ELSE 0
                END
            ) as passed_ct,
            SUM(
                CASE
                    WHEN COALESCE(latest_results.disposition, 'Confirmed') = 'Confirmed'
                    AND latest_results.result_status = 'Warning' THEN 1
                    ELSE 0
                END
            ) as warning_ct,
            SUM(
                CASE
                    WHEN COALESCE(latest_results.disposition, 'Confirmed') = 'Confirmed'
                    AND latest_results.result_status = 'Failed' THEN 1
                    ELSE 0
                END
            ) as failed_ct,
            SUM(
                CASE
                    WHEN COALESCE(latest_results.disposition, 'Confirmed') = 'Confirmed'
                    AND latest_results.result_status = 'Error' THEN 1
                    ELSE 0
                END
            ) as error_ct,
            SUM(
                CASE
                    WHEN COALESCE(latest_results.disposition, 'Confirmed') IN ('Dismissed', 'Inactive') THEN 1
                    ELSE 0
                END
            ) as dismissed_ct
        FROM {schema}.test_suites suites
            LEFT JOIN {schema}.test_runs latest_run ON (
                suites.last_complete_test_run_id = latest_run.id
            )
            LEFT JOIN {schema}.test_results latest_results ON (
                latest_run.id = latest_results.test_run_id
            )
        GROUP BY suites.table_groups_id
    )
    SELECT groups.id::VARCHAR(50),
        groups.table_groups_name,
        latest_profile.id as latest_profile_id,
        latest_profile.profiling_starttime as latest_profile_start,
        latest_profile.table_ct as latest_profile_table_ct,
        latest_profile.column_ct as latest_profile_column_ct,
        latest_profile.anomaly_ct as latest_anomalies_ct,
        latest_profile.definite_ct as latest_anomalies_definite_ct,
        latest_profile.likely_ct as latest_anomalies_likely_ct,
        latest_profile.possible_ct as latest_anomalies_possible_ct,
        latest_profile.dismissed_ct as latest_anomalies_dismissed_ct,
        latest_tests.test_starttime as latest_tests_start,
        latest_tests.test_suite_ct as latest_tests_suite_ct,
        latest_tests.test_ct as latest_tests_ct,
        latest_tests.passed_ct as latest_tests_passed_ct,
        latest_tests.warning_ct as latest_tests_warning_ct,
        latest_tests.failed_ct as latest_tests_failed_ct,
        latest_tests.error_ct as latest_tests_error_ct,
        latest_tests.dismissed_ct as latest_tests_dismissed_ct
    FROM {schema}.table_groups as groups
        LEFT JOIN latest_profile ON (groups.id = latest_profile.table_groups_id)
        LEFT JOIN latest_tests ON (groups.id = latest_tests.table_groups_id)
    WHERE groups.project_code = '{project_code}';
    """
    return db.retrieve_data(sql)

File: testgen/ui/components/frontend/js/pages/overview.js
/**
 * @typedef ProjectSummary
 * @type {object}
 * @property {number} table_groups_count
 * @property {number} test_suites_count
 * @property {number} test_definitions_count
 * @property {number} test_runs_count
 * @property {number} profiling_runs_count
 * @property {number} connections_count
 * @property {string} default_connection_id
 * 
 * @typedef TestSuiteSummary
 * @type {object}
 * @property {string} id
 * @property {string} test_suite
 * @property {number} test_ct
 * @property {string} latest_auto_gen_date
 * @property {string} latest_run_start
 * @property {string} latest_run_id
 * @property {number} last_run_test_ct
 * @property {number} last_run_passed_ct
 * @property {number} last_run_warning_ct
 * @property {number} last_run_failed_ct
 * @property {number} last_run_error_ct
 * @property {number} last_run_dismissed_ct
 * 
 * @typedef TableGroupSummary
 * @type {object}
 * @property {string} id
 * @property {string} table_groups_name
 * @property {string} table_groups_name
 * @property {number?} dq_score
 * @property {number?} dq_score_profiling
 * @property {number?} dq_score_testing
 * @property {string} latest_profile_id
 * @property {string} latest_profile_start
 * @property {number} latest_profile_table_ct
 * @property {number} latest_profile_column_ct
 * @property {number} latest_anomalies_ct
 * @property {number} latest_anomalies_definite_ct
 * @property {number} latest_anomalies_likely_ct
 * @property {number} latest_anomalies_possible_ct
 * @property {number} latest_anomalies_dismissed_ct
 * @property {string} latest_tests_start
 * @property {number} latest_tests_suite_ct
 * @property {number} latest_tests_ct
 * @property {number} latest_tests_passed_ct
 * @property {number} latest_tests_warning_ct
 * @property {number} latest_tests_failed_ct
 * @property {number} latest_tests_error_ct
 * @property {number} latest_tests_dismissed_ct
 * @property {TestSuiteSummary[]} test_suites
 * @property {boolean} expanded
 * 
 * @typedef SortOption
 * @type {object}
 * @property {string} value
 * @property {string} label
 * @property {boolean} selected
 * 
 * @typedef Properties
 * @type {object}
 * @property {ProjectSummary} project
 * @property {TableGroupSummary[]} table_groups
 * @property {SortOption[]} table_groups_sort_options
 */
import van from '../van.min.js';
import { Streamlit } from '../streamlit.js';
import { emitEvent, getValue, loadStylesheet, truncateFloat } from '../utils.js';
import { formatTimestamp } from '../display_utils.js';
import { Card } from '../components/card.js';
import { Caption } from '../components/caption.js';
import { ExpanderToggle } from '../components/expander_toggle.js';
import { Select } from '../components/select.js';
import { Input } from '../components/input.js';
import { Link } from '../components/link.js';
import { SummaryBar } from '../components/summary_bar.js';
import { EmptyState } from '../components/empty_state.js';
import { Metric } from '../components/metric.js';

const { div, h3, hr, span, strong } = van.tags;

const Overview = (/** @type Properties */ props) => {
    loadStylesheet('overview', stylesheet);
    Streamlit.setFrameHeight(1);
    window.frameElement.style.setProperty('height', 'calc(100vh - 200px)');
    window.testgen.isPage = true;

    const isEmpty = van.derive(() => {
        const projectSummary = getValue(props.project)
        return projectSummary.test_runs_count <= 0 && projectSummary.profiling_runs_count <= 0;
    });
    const tableGroups = van.derive(() => getValue(props.table_groups));
    const filteredTableGroups = van.state(getValue(tableGroups));
    const tableGroupsSearchTerm = van.state('');
    const tableGroupsSortOption = van.state(getValue(props.table_groups_sort_options).find(o => o.selected)[0]?.value);
    const sortFunctions = {
        table_groups_name: (a, b) => a.table_groups_name.toLowerCase().localeCompare(b.toLowerCase()),
        latest_activity_date: (a, b) => Math.max(b.latest_profile_start, b.latest_tests_start) - Math.max(a.latest_profile_start, a.latest_tests_start),
    };

    van.derive(() => {
        const searchTerm = getValue(tableGroupsSearchTerm);
        filteredTableGroups.val = getValue(tableGroups).filter(group => group.table_groups_name.toLowerCase().includes(searchTerm.toLowerCase() ?? ''));
    });

    van.derive(() => {
        const sortByField = getValue(tableGroupsSortOption);
        const sortFn = sortFunctions[sortByField] ?? sortFunctions.latest_activity_date
        filteredTableGroups.val = getValue(filteredTableGroups).sort(sortFn);
    });

    return div(
        { class: 'flex-column tg-overview' },
        () => !getValue(isEmpty)
            ? div(
                { class: 'flex-row fx-align-stretch', style: 'gap: 1rem;' },
                Card({
                    id: 'overview-project-summary',
                    class: 'tg-overview--project',
                    content: [
                        () => div(
                            { class: 'flex-row fx-align-flex-start' },
                            // div(
                            //     { class: 'flex-column tg-overview--project--score' },
                            //     Caption({ content: 'Project HIT score', style: 'margin-bottom: 16px;' }),
                            //     Metric({value: 100, delta: 10}),
                            // ),
                            () => {
                                return div(
                                    { class: 'flex-column tg-overview--project--summary' },
                                    Caption({content: 'Project Summary', style: 'margin-bottom: 16px;' }),
                                    div(
                                        { style: 'margin-bottom: 8px;' },
                                        strong({ style: 'margin-right: 4px;' }, props.project.val.table_groups_count),
                                        span('table groups'),
                                    ),
                                    div(
                                        { style: 'margin-bottom: 8px;' },
                                        strong({ style: 'margin-right: 4px;' }, props.project.val.test_suites_count),
                                        span('test suites'),
                                    ),
                                    div(
                                        { style: 'margin-bottom: 8px;' },
                                        strong({ style: 'margin-right: 4px;' }, props.project.val.test_definitions_count),
                                        span('test definitions'),
                                    ),
                                );
                            }
                        ),
                    ],
                }),
            )
            : ConditionalEmptyState(getValue(props.project)),
        () => !getValue(isEmpty)
            ? div(
                { class: 'flex-row' },
                h3(() => `Table Groups (${tableGroups?.val?.length ?? 0})`),
                span({ style: 'margin-right: auto;' }),
                Input({
                    label: 'Search by table group name',
                    width: 230,
                    height: 38,
                    style: 'font-size: 14px;',
                    onChange: (value) => tableGroupsSearchTerm.val = value,
                }),
                span({ style: 'margin-right: 1rem;' }),
                Select({
                    label: 'Sort by',
                    options: props.table_groups_sort_options?.val ?? [],
                    height: 38,
                    style: 'font-size: 14px;',
                    onChange: (value) => tableGroupsSortOption.val = value,
                }),
            )
            : undefined,
        () => !getValue(isEmpty)
            ? div(
                { class: 'flex-column mt-2' },
                getValue(filteredTableGroups).map(tableGroup => TableGroupCard(tableGroup)),
            )
            : undefined,
    );
}

const TableGroupCard = (/** @type TableGroupSummary */ tableGroup) => {
    return Card({
        title: () => div(
            { class: 'flex-row' },
            span({ style: 'display: block; margin-right: auto;' }, tableGroup.table_groups_name),
            ExpanderToggle({
                default: tableGroup.expanded,
                style: 'font-size: 14px !important; font-weight: 400;',
                onExpand: () => {
                    emitEvent('TableGroupExpanded', {payload: tableGroup.id});
                },
                onCollapse: () => {
                    emitEvent('TableGroupCollapsed', {payload: tableGroup.id});
                },
            }),
        ),
        class: 'tg-overview--table-group-card',
        content: () => div(
            { class: 'flex-column' },
            div(
                { class: 'flex-row flex-align-flex-start' },
                div(
                    { class: 'flex-column fx-flex' },
                    TableGroupLatestProfile(tableGroup),
                ),
                div(
                    { class: 'flex-column fx-flex' },
                    TableGroupLatestTestResults(tableGroup),
                ),
                div(
                    { class: 'flex-column fx-align-flex-center', style: 'flex: 0 1 10%;' },
                    Metric({ value: tableGroup.dq_score ?? '--' }),
                    Caption({ content: 'Score' }),
                ),
            ),
            tableGroup.expanded
                ? hr({ class: 'tg-overview--table-group-divider' })
                : undefined,
            tableGroup.expanded
                ? TableGroupTestSuiteSummary(tableGroup.test_suites)
                : undefined,
        )
    });
};

const TableGroupLatestProfile = (/** @type TableGroupSummary */ tableGroup) => {
    return [
        Caption({ content: 'Latest profile' }),
        div(
            { class: 'flex-row mb-3' },
            Link({
                label: formatTimestamp(tableGroup.latest_profile_start),
                href: 'profiling-runs:results',
                params: { run_id: tableGroup.latest_profile_id },
            }),
            span({ class: 'mr-1 ml-1' }, '|'),
            span(`Profiling score: ${tableGroup.dq_score_profiling}`),
        ),
        div(
            { class: 'flex-row mb-3' },
            strong({ class: 'mr-1' }, tableGroup.latest_profile_table_ct),
            span('tables'),
            span({ class: 'mr-1 ml-1' }, '|'),
            strong({ class: 'mr-1' }, tableGroup.latest_profile_column_ct),
            span('columns'),
            span({ class: 'mr-1 ml-1' }, '|'),
            Link({
                label: `${tableGroup.latest_anomalies_ct} hygiene issues`,
                href: 'profiling-runs:hygiene',
                params: {
                    run_id: tableGroup.latest_profile_id,
                },
                width: 150,
            })
        ),
        () => tableGroup.latest_anomalies_ct
            ? SummaryBar({
                items: [
                    { label: 'Definite', value: parseInt(tableGroup.latest_anomalies_definite_ct), color: 'red' },
                    { label: 'Likely', value: parseInt(tableGroup.latest_anomalies_likely_ct), color: 'orange' },
                    { label: 'Possible', value: parseInt(tableGroup.latest_anomalies_possible_ct), color: 'yellow' },
                    { label: 'Dismissed', value: parseInt(tableGroup.latest_anomalies_dismissed_ct), color: 'grey' },
                ],
                height: 12,
                width: 280,
            })
            : '',
    ];
};

const TableGroupLatestTestResults = (/** @type TableGroupSummary */ tableGroup) => {
    return [
        Caption({ content: 'Latest test results' }),
        () => tableGroup.latest_tests_ct
            ? div(
                { class: 'flex-column' },
                span(
                    { class: 'mb-3' },
                    `${truncateFloat(tableGroup.latest_tests_passed_ct * 100 / tableGroup.latest_tests_ct)}% passed | Test score: ${tableGroup.dq_score_testing}`,
                ),
                div(
                    { class: 'flex-row mb-3' },
                    strong({ class: 'mr-1' }, tableGroup.latest_tests_ct),
                    span({ class: 'mr-1' }, 'tests in'),
                    strong({ class: 'mr-1' }, tableGroup.latest_tests_suite_ct),
                    span('test suites'),
                ),
                SummaryBar({
                    items: [
                        { label: 'Passed', value: parseInt(tableGroup.latest_tests_passed_ct), color: 'green' },
                        { label: 'Warning', value: parseInt(tableGroup.latest_tests_warning_ct), color: 'yellow' },
                        { label: 'Failed', value: parseInt(tableGroup.latest_tests_failed_ct), color: 'red' },
                        { label: 'Error', value: parseInt(tableGroup.latest_tests_error_ct), color: 'brown' },
                        { label: 'Dismissed', value: parseInt(tableGroup.latest_tests_dismissed_ct), color: 'grey' },
                    ],
                    height: 12,
                    width: 350,
                })
            )
            : span('--'),
    ];
};

const TableGroupTestSuiteSummary = (/** @type TestSuiteSummary[] */testSuites) => {
    return div(
        { class: 'flex-column' },
        div(
            { class: 'flex-row mb-4' },
            Caption({ content: 'Test Suite', style: 'flex: 1 1 20%;' }),
            Caption({ content: 'Latest Generation', style: 'flex: 1 1 15%;' }),
            Caption({ content: 'Latest Run', style: 'flex: 1 1 15%;' }),
            Caption({ content: 'Latest Results', style: 'flex: 1 1 50%;' }),
        ),
        testSuites.map(suite => div(
            { class: 'flex-row' },
            div(
                { class: 'flex-column', style: 'flex: 1 1 20%;' },
                Link({
                    label: suite.test_suite,
                    href: 'test-suites:definitions',
                    params: { test_suite_id: suite.id },
                }),
                Caption({ content: `${suite.test_ct} tests`}),
            ),
            suite.latest_auto_gen_date
                ? span(
                    { style: 'flex: 1 1 15%;' },
                    formatTimestamp(suite.latest_auto_gen_date),
                )
                : '--',
            suite.latest_run_id
                ? Link({
                    label: formatTimestamp(suite.latest_auto_gen_date),
                    href: 'test-runs:results',
                    params: { run_id: suite.latest_run_id },
                    style: 'flex: 1 1 15%;',
                })
                : '--',
            suite.last_run_test_ct
                ? div(
                    { style: 'flex: 1 1 50%;' },
                    SummaryBar({
                        items: [
                            { label: 'Passed', 'value': parseInt(suite.last_run_passed_ct), color: 'green' },
                            { label: 'Warning', 'value': parseInt(suite.last_run_warning_ct), color: 'yellow' },
                            { label: 'Failed', 'value': parseInt(suite.last_run_failed_ct), color: 'red' },
                            { label: 'Error', 'value': parseInt(suite.last_run_error_ct), color: 'brown' },
                            { label: 'Dismissed', 'value': parseInt(suite.last_run_dismissed_ct), color: 'grey' },
                        ],
                        width: 200,
                        height: 8,
                    })
                )
                : '--',
        ))
    );
};

const ConditionalEmptyState = (/** @type ProjectSummary */ project) => {
    const forConnections = {
        message: {
            line1: 'Begin by connecting your database.',
            line2: 'TestGen delivers data quality through data profiling, hygiene review, test generation, and test execution.',
        },
        link: {
            label: 'Go to Connections',
            href: 'connections',
        },
    };
    const forTablegroups = {
        message: {
            line1: 'Profile your tables to detect hygiene issues',
            line2: 'Create table groups for your connected databases to run data profiling and hygiene review.',
        },
        link: {
            label: 'Go to Table Groups',
            href: 'connections:table-groups',
            params: { connection_id: project.default_connection_id },
        },
    };

    const args = project.connections_count > 0 ? forTablegroups : forConnections;

    return EmptyState({
        icon: 'home',
        label: 'Your project is empty',
        ...args,
    });
};

const stylesheet = new CSSStyleSheet();
stylesheet.replace(`
.tg-overview {
    width: 100%;
}

.tg-overview--project {
    border: 1px solid var(--border-color);
    width: 50%;
}

.tg-overview--project--score {
    margin-right: auto;
}

.tg-overview--project--summary {
    margin-right: auto;
}

.tg-overview--table-group-card {
    border: 1px solid var(--border-color);
}

hr.tg-overview--table-group-divider {
    height: 1px;
    margin: 8px 0 12px;
    padding: 0px;
    color: inherit;
    background-color: transparent;
    border-top: none;
    border-right: none;
    border-left: none;
    border-image: initial;
    border-bottom: 1px solid rgba(49, 51, 63, 0.2);
}
`);

export { Overview };
